### PR TITLE
hw: adding tpm tis spi (not using cs)

### PR DESCRIPTION
**Pull request by chen-iris**
Signed-off-by: Iris Chen <irischenlj@fb.com>

- Not using CS
`SPI TPMv2.0 found (vid:1014, did:0001, rid:01)`
P514156885

### CODE REVIEW

File: hw/tpm/tpm_tis_spi.c [hunk 1]
```diff
@@ -0,0 +1,254 @@
...
+            tts->addr.bytes[--tts->addr_idx] = (uint8_t)tx;
+
+            if (tts->addr_idx == 0) {
+                if (tts->first_byte.rwflag == TIS_SPI_PKT_STATE_DATA_WR)
+                tts->tpm_state = TIS_SPI_PKT_STATE_DATA_WR;
+            } else { /* read and get the data ready */
+                tts->data = (uint32_t)tpm_tis_mmio_read(&tts->tpm_state, tts->addr.addr,
+                                      tts->first_byte.data_expected_size);
+                tts->tpm_state = TIS_SPI_PKT_STATE_DATA_RD;
+                }
```
Comment (zhdaniel12) [inline hw/tpm/tpm_tis_spi.c]:
```suggestion
                if (tts->first_byte.rwflag == TIS_SPI_PKT_STATE_DATA_WR)
                    tts->tpm_state = TIS_SPI_PKT_STATE_DATA_WR;
                } else { /* read and get the data ready */
                    tts->data = (uint32_t)tpm_tis_mmio_read(&tts->tpm_state, tts->addr.addr,
                                      tts->first_byte.data_expected_size);
                    tts->tpm_state = TIS_SPI_PKT_STATE_DATA_RD;
                }
```

File: hw/tpm/tpm_tis_spi.c [hunk 2]
```diff
@@ -0,0 +1,254 @@
...
+                 tts->spi_state = TIS_SPI_PKT_STATE_DONE;
+            }
+            break;
+        case TIS_SPI_PKT_STATE_DATA_RD:
+            r = tts->data.bytes[tts->data_idx++];
+            if (tts->data_idx == tts->first_byte.data_expected_size) {
+                tts->spi_state = TIS_SPI_PKT_STATE_DONE;
+            }
+            break;
+        case TIS_SPI_PKT_STATE_START:
```
Comment (zhdaniel12) [inline hw/tpm/tpm_tis_spi.c]: let's put this state in the first part of this switch.

**Review by zhdaniel12 — CHANGES_REQUESTED** [review]
you need add kconfig and update the meson.build to enable the build.

File: hw/tpm/tpm_tis_spi.c [hunk 3]
```diff
@@ -0,0 +1,253 @@
...
+
+static uint32_t tpm_tis_spi_transfer8(SSIPeripheral *ss, uint32_t tx)
+{
+    TPMSpiState *tts = TPM_TIS_SPI(ss);
+    uint32_t r = 1;
+
+    switch (tts->tis_spi_state) {
+    case TIS_SPI_PKT_STATE_START:
+        tts->first_byte.byte = (uint8_t)tx;
+        tts->first_byte.data_expected_size++;
```
Comment (zhdaniel12) [inline hw/tpm/tpm_tis_spi.c]: here, I think we have a small bug. as the bit filed data_expected_size only 6 bit, but it need to represent size: 1 ~ 64, thus it use 0 based size. 
To make our code simple, suggest you add a new member in class TpmSpiState, as uint32_t data_size. And here 
```suggestion
        tts->data_size = first_byte.data_expected_size + 1;
```
And all the other code place use tts->data_size.

File: hw/tpm/tpm_tis_spi.c [hunk 4]
```diff
@@ -0,0 +1,253 @@
...
+
+    return r;
+}
+
+static uint32_t tpm_tis_spi_transfer8_ex(SSIPeripheral *ss, uint32_t tx)
+{
+    TPMSpiState *tts = TPM_TIS_SPI(ss);
+    SSIBus *ssibus = (SSIBus *)qdev_get_parent_bus(DEVICE(tts));
+
+    uint8_t prev_state = tts->tis_spi_state;
```
Comment (zhdaniel12) [inline hw/tpm/tpm_tis_spi.c]: you can use TisSpiPktState here

File: hw/tpm/tpm_tis_spi.c [hunk 5]
```diff
@@ -0,0 +1,253 @@
...
+    uint32_t r = tpm_tis_spi_transfer8(ss, tx);
+    uint8_t curr_state = tts->tis_spi_state;
```
Comment (zhdaniel12) [inline hw/tpm/tpm_tis_spi.c]: here also

File: hw/tpm/tpm_tis_spi.c [hunk 6]
```diff
@@ -0,0 +1,253 @@
...
+
+    if (ssibus->preread &&
+       /* cmd state has changed into DATA reading state */
```
Comment (zhdaniel12) [inline hw/tpm/tpm_tis_spi.c:129]: strong suggest add comments here explain our preread hacking. 
give example in my previous comments.

**Review by zhdaniel12 — CHANGES_REQUESTED** [review]
1. fix that expected_size bug
2. add comments explain our preread hacking
3. shall we have all the variable, class, instance, functions naming consistently start with TpmTisSpi or tpm_tis_spi ( i.e. TpmSpiState ==> TpmTisSpiState )? like a name space

File: hw/tpm/tpm_tis_spi.c (hunk 3 above)
Comment (chen-iris) [reply hw/tpm/tpm_tis_spi.c]: This is a good catch. Thanks!

**Comment by chen-iris** [comment]
> 1. fix that expected_size bug
> 2. add comments explain our preread hacking
> 3. shall we have all the variable, class, instance, functions naming consistently start with TpmTisSpi or tpm_tis_spi ( i.e. TpmSpiState ==> TpmTisSpiState )? like a name space

I have addressed all of these comments. 

For 3, I have tried to make all variable, class, instance, and function names start with this prefix. Let me know if you have further suggestions.

**Review by zhdaniel12 — APPROVED** [review]

**Pull request merged**